libfeed/feed-store-provider.c

```diff
@@ -0,0 +1,1074 @@
+                                                     bundle->quote_proxy,
+                                                     cancellable,
+                                                     individual_task_result_completed,
+                                                     individual_task_result_closure_new (all_tasks_results_closure_allocate_new_result_slot (all_tasks_closure),
+                                                                                         all_tasks_closure));
+    }
+}
+
+static void
+received_all_unordered_card_array_results_from_queries (GObject      *source,
+                                                        GAsyncResult *result,
+                                                        gpointer      user_data)
+{
+  g_autoptr(GError) local_error = NULL;
+  g_autoptr(GPtrArray) results = g_task_propagate_pointer (G_TASK (result),
+                                                           &local_error);
+  g_autoptr(GPtrArray) result_arrays = g_ptr_array_new_with_free_func ((GDestroyNotify) g_ptr_array_unref);
+  g_autoptr(GPtrArray) flat_mapped_array = NULL;
+  g_autoptr(GTask) task = user_data;
+  guint i = 0;
+
+  /* This basically shouldn't happen, but handle it anyway */
+  if (result == NULL)
+    {
+      g_message ("Error getting all unordered card results: %s", local_error->message);
+      g_task_return_error (task, local_error);
+      return;
+    }
+
+  /* Go through each of the results, complain about the ones that failed
+   * but keep the ones that yielded some models. We'll flat-map them all
+   * together later */
+  for (i = 0; i < results->len; ++i)
+    {
+      g_autoptr(GPtrArray) result_array = g_task_propagate_pointer (g_ptr_array_index (results, i),
+                                                                    &local_error);
+
+      if (result_array == NULL)
+        {
+          g_message ("Query failed: %s", local_error->message);
+          g_clear_error (&local_error);
+          continue;
+        }
+
+      g_ptr_array_add (result_arrays, g_steal_pointer (&result_array));
+    }
+
+  flat_mapped_array = flat_map_ptr_arrays (result_arrays,
+                                           (ElementCopyFunc) g_object_ref,
+                                           g_object_unref);
+  g_task_return_pointer (task,
+                         g_steal_pointer (&flat_mapped_array),
+                         (GDestroyNotify) g_ptr_array_unref);
+}
+
+static GPtrArray *
+pointerv_to_g_ptr_array (gpointer *pointerv)
+{
+  gpointer *iter = pointerv;
+  GPtrArray *ptr_array = g_ptr_array_new ();
+
+  for (; *iter != NULL; ++iter)
+    g_ptr_array_add (ptr_array, *iter);
+
+  return ptr_array;
+}
+
+/**
+ * eos_discovery_feed_unordered_results_from_queries_finish:
+ * @result: A #GAsyncResult
+ * @error: A #GError
+ *
+ * Complete the call to eos_discovery_feed_unordered_results_from_queries.
+ *
+ * Returns: (transfer container) (element-type EosDiscoveryFeedBaseCardStore):
```

Comment:
It does not look like the `(transfer container)` is right here; all the objects inside the array are owned by the array itself, no?

Reply:
I traced this down to https://gitlab.gnome.org/GNOME/gjs/blob/master/gi/arg.cpp#L3457  - it looks like gjs (and pygi, too!) will free all the elements unless using transfer-container. Not sure why it doesn't handle transfer-full